src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CompanyAccountServiceImpl.java

```diff
@@ -1,33 +1,98 @@
 package uk.gov.companieshouse.api.accounts.service.impl;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
-import uk.gov.companieshouse.api.accounts.Kind;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+import uk.gov.companieshouse.api.accounts.LinkType;
 import uk.gov.companieshouse.api.accounts.model.entity.CompanyAccountEntity;
 import uk.gov.companieshouse.api.accounts.model.rest.CompanyAccount;
 import uk.gov.companieshouse.api.accounts.service.CompanyAccountService;
+import uk.gov.companieshouse.api.accounts.transaction.Transaction;
+import uk.gov.companieshouse.api.accounts.transaction.TransactionManager;
 import uk.gov.companieshouse.api.accounts.transformer.GenericTransformer;
 
 @Service
 public class CompanyAccountServiceImpl extends
         AbstractServiceImpl<CompanyAccount, CompanyAccountEntity> implements CompanyAccountService {
 
+    @Autowired
+    private TransactionManager transactionManager;
+
     @Autowired
     public CompanyAccountServiceImpl(
             @Qualifier("companyAccountRepository") MongoRepository<CompanyAccountEntity, String> mongoRepository,
             @Qualifier("companyAccountTransformer") GenericTransformer<CompanyAccount, CompanyAccountEntity> transformer) {
         super(mongoRepository, transformer);
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public CompanyAccount createCompanyAccount(CompanyAccount companyAccount) {
+        HttpServletRequest request = getRequestFromContext();
+        ResponseEntity<Transaction> transaction = getTransactionFromSession(request);
+
+        addKind(companyAccount);
+        addEtag(companyAccount);
+
+        CompanyAccountEntity companyAccountEntity = genericTransformer.transform(companyAccount);
+
+        addEntityID(companyAccountEntity);
+        addEntityLinks(companyAccountEntity, transaction);
+
+        mongoRepository.insert(companyAccountEntity);
+
+        transactionManager.updateTransaction(transaction.getBody().getId(), request.getHeader("X-Request-Id"), getCompanyAccountSelfLink(companyAccountEntity));
```

Comment:
This could be simplified to:
```java
String companyAccountsLink = getCompanyAccountsLink(transaction, entity.getId());
addLinks(companyAccountEntity);
mongoRepository.insert(companyAccountEntity);
transactionManager.updateTransaction(transaction.getBody().getId(), request.getHeader("X-Request-Id"), companyAccountsLink);
```

Reply:
I can't use the overriden addLinks() method because of the way the abstract service is set up. I will have to change the abstract service, which will break the small full stuff. Additionally, the entityLinks has to be after the setting of the id of the company-accounts resource

Reply:
The addLinks() is my fault I wanted to get some early visibility of the sort of methods we would need to implement, I have a branch of code where I revisit that due for PR imminently. I do agree with Jo's simplification of the code though that makes the last call to genericTransformer.transform(companyAccountEntity) unnecessary.